### DISCUSSION TITLE
Does Devito support any methods to reduce GPU memory usage?

### DISCUSSION THREAD

**Discussion by jinshanmu** [devitocodes/devito#2427] [Q&A] [answered]
The computation task requires ~100GB memory. My CPU has 256GB, but the GPU only has 16GB. Is it possible to load a smaller memory to the GPU once at a time and keep the rest at the CPU so that the GPU can be utilized at full extent without being out of memory?

**Comment by mloubout** [comment]
What type of task is it? Is that inversion, i.e computing a gradient and saving the forward wavefield, or simple forward simulation of a very large model?

**Reply by jinshanmu** [reply]
The breast3D example from Stride. A large 3D FWI case.

**Reply by mloubout** [reply]
Then you have a few options in devito, including time subsampling, optimal checkpointing, and streaming. However, I am not sure if those are exposed in Stride to the user. (@ccuetom )

**Reply by ccuetom** [reply]
@jinshanmu Time subsampling and streaming (with DevitoPRO) are exposed to the user to control. Optimal checkpointing is not. You can also consider tweaking the sampling and number of timesteps of the model you are running.

**Reply by jinshanmu** [reply]
> @jinshanmu Time subsampling and streaming (with DevitoPRO) are exposed to the user to control. Optimal checkpointing is not. You can also consider tweaking the sampling and number of timesteps of the model you are running.

If I understand correctly, is time subsampling the only way currently to achieve this on the original example using Stride with Devito?

**Reply by ccuetom** [reply]
It is the only way in Devito Open-Source, other than changing the grid of the problem

**Reply by jinshanmu** [reply]
I see. Btw, are there example codes for time subsampling to optimize GPU that I can learn from? I searched in the tutorials of both Devito and Stride and did not find any, only time subsampling for snapshots in Devito tutorials it seems. Thank you! @mloubout @ccuetom

**Reply by ccuetom** [reply, marked as answer]
for subsampling you only need to pass `save_undersampling=<integer subsampling>` to the `adjoint` call

**Reply by jinshanmu** [reply]
> for subsampling you only need to pass `save_undersampling=<integer subsampling>` to the `adjoint` call

Much better performance now! Thx!
